<launch>
  <arg name="uav_name" default="$(optenv UAV_NAME uav2)"/>

  <arg name="standalone" default="false"/>

  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg uav_name)_uvdar_nodelet_manager"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>

  <arg name="threshold" default="50"/>

  <arg name="left" default="$(optenv BLUEFOX_UV_LEFT)"/>
  <arg name="right" default="$(optenv BLUEFOX_UV_RIGHT)"/>
  <arg name="left_camera_name" default="mv_$(arg left)"/>
  <arg name="right_camera_name" default="mv_$(arg right)"/>
  <arg name="expose_us_left" default="$(optenv BLUEFOX_UV_LEFT_EXPOSE_US)"/>
  <arg name="expose_us_right" default="$(optenv BLUEFOX_UV_RIGHT_EXPOSE_US)"/>

  <arg name="calibrations_folder" default="$(find mrs_uav_general)/config/uvdar_calibrations"/>

  <arg name="blink_process_rate" default="10"/>

  <arg name="accumulator_length"         default="18"/>
  <arg name="pitch_steps"         default="16"/>
  <arg name="yaw_steps"         default="16"/>
  <arg name="max_pixel_shift"         default="4"/>

  <arg name="filterDistLength"         default="10"/>
  <arg name="filterOrientationLength"         default="10"/>

  <!-- Node Settings -->
  <arg name="output" default="screen"/>
  <arg name="proc" default="false"/>
  <arg name="view" default="false"/>
  <arg name="calib" default="false"/>

  <arg name="profiling" default="false"/>
  <arg name="debug" default="false"/>
  <arg name="visual_debug" default="false"/>
  <arg name="gui" default="true"/>
  <!-- <arg name="gui" default="false"/> -->

  <arg name="publish" default="true"/>
  <arg name="useOdom" default="false"/>
  <arg name="cemeraRotated" default="false"/>
  <arg name="TimeScale" default="1.0"/>

  <!-- <arg name="frameMRSid" default="$(optenv MRS_ID)"/> -->

  <arg name="beacon" default="false"/>

  <!-- <arg name="frequency1" value="6"/> -->
  <!-- <arg name="frequency2" value="15"/> -->
  <!-- <arg name="frequency3" value="30"/> -->
  <arg name="id1" value="0"/>
  <arg name="id2" value="1"/>
  <arg name="id3" value="2"/>
  <arg name="id4" value="3"/>
  <arg name="id5" value="4"/>
  <arg name="id6" value="5"/>
  <arg name="id7" value="6"/>
  <arg name="id8" value="7"/>
  <arg name="id9" value="8"/>
  <arg name="id10" value="9"/>
  <arg name="id11" value="10"/>
  <arg name="id12" value="11"/>

  <arg name="sequence_file" default="$(find uvdar_core)/config/selected.txt"/>
  <arg name="model_file" default="$(find uvdar_core)/config/models/quadrotor_foursided_px4_motor_aligned.txt"/>
  <!-- ideal angle: 1.221730476 - pi/2 -->

      <!-- <origin xyz="0.108 0 0.085" rpy="1.57079632679 4.71238898038 1.57079632679" /> -->
  <!--Usage: static_transform_publisher x y z yaw pitch roll frame_id child_frame_id  period(milliseconds)-->
<!-- OR -->
  <group ns="$(arg uav_name)">

  <node
    name="uvcam_left_tf_$(arg uav_name)"
    pkg="tf2_ros"
    type="static_transform_publisher"
    args="0.03  0.10 0.06 -0.3490658504 0.0 -1.57079632679 $(arg uav_name)/fcu $(arg uav_name)/uvcam_left"/>

  <node
    name="uvcam_right_tf_$(arg uav_name)"
    pkg="tf2_ros"
    type="static_transform_publisher"
    args="0.03 -0.10 0.06 -2.792526803 0.0 -1.57079632679 $(arg uav_name)/fcu $(arg uav_name)/uvcam_right"/>

  <node
    name="uvcam_back_tf_$(arg uav_name)"
    pkg="tf2_ros"
    type="static_transform_publisher"
    args="-0.10 0.0 -0.02 0.0 1.57079632679 3.14159265 $(arg uav_name)/fcu $(arg uav_name)/uvcam_back"/>

    <!-- <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" > -->
    <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" launch-prefix="debug_roslaunch">
      <param name="num_worker_threads" value="8" />
    </node>

    <node name="bluefox_emulator" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBluefoxEmulator $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="bash -c 'sleep 5; $0 $@'">
      <rosparam param="camera_output_topics"> ["camera_left", "camera_right", "camera_back"] </rosparam>
      <rosparam param="calib_files"> ["default", "default", "default"] </rosparam>
      <remap from="~camera_left" to="/$(arg uav_name)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right" to="/$(arg uav_name)/uvdar_bluefox_right/image_raw"/>
      <remap from="~camera_back" to="/$(arg uav_name)/uvdar_bluefox_back/image_raw"/>
      <remap from="~camera_left_transfer" to="/gazebo/$(arg uav_name)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right_transfer" to="/gazebo/$(arg uav_name)/uvdar_bluefox_right/image_raw"/>
      <remap from="~camera_back_transfer" to="/gazebo/$(arg uav_name)/uvdar_bluefox_back/image_raw"/>
    </node>



    <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="false"> 
    <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="debug_roslaunch"> --> 
      <!-- <node name="uv_detect" pkg="uvdar" type="uv_detector_node" output="screen" launch-prefix="urxvt -e gdb -q -x /home/viktor/gdb.cmds -/-args"> -->
        <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="valgrind -/-tool=callgrind -/-callgrind-out-file=/home/viktor/callgrind.out -/-instr-atstart=no -/-collect-atstart=yes"> --> 

      <param name="uav_name"		type="string" value="$(arg uav_name)"/> 
      <param name="debug"		type="bool" value="$(arg debug)"/> 
      <!-- <param name="gui"             type="bool" value="true"/> -->
      <param name="gui"             type="bool" value="false"/>
      <!-- <param name="publish_visualization"             type="bool" value="true"/> -->
      <param name="publish_visualization"             type="bool" value="false"/>
      <param name="publish"         type="bool" value="$(arg publish)"/>
      <param name="justReport" type="bool" value="true"/>
      <param name="threshold" type="int" value="$(arg threshold)"/>

      <rosparam param="camera_topics"> ["camera_left", "camera_right", "camera_back"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_left", "points_seen_right", "points_seen_back"] </rosparam>

      <!-- <param name="use_masks"         type="bool" value="false"/> -->
      <param name="use_masks"         type="bool" value="true"/>
      <rosparam param="mask_file_names" subst_value="True"> ["$(arg calibrations_folder)/masks/blank_bluefox.png", "$(arg calibrations_folder)/masks/blank_bluefox.png", "$(arg calibrations_folder)/masks/blank_bluefox.png"] </rosparam>

      <param name="publish_sun_points"         type="bool" value="true"/>

      <remap from="~camera_left" to="/$(arg uav_name)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right" to="/$(arg uav_name)/uvdar_bluefox_right/image_raw"/>
      <remap from="~camera_back" to="/$(arg uav_name)/uvdar_bluefox_back/image_raw"/>
      <remap from="~points_seen_left" to="/$(arg uav_name)/uvdar/points_seen_left"/>
      <remap from="~points_seen_right" to="/$(arg uav_name)/uvdar/points_seen_right"/>
      <remap from="~points_seen_back" to="/$(arg uav_name)/uvdar/points_seen_back"/>
      <remap from="~points_seen_left/sun" to="/$(arg uav_name)/uvdar/points_seen_left/sun"/>
      <remap from="~points_seen_right/sun" to="/$(arg uav_name)/uvdar/points_seen_right/sun"/>
      <remap from="~points_seen_back/sun" to="/$(arg uav_name)/uvdar/points_seen_back/sun"/>

      <remap from="~odometry" to="/$(arg uav_name)/mrs_odometry/new_odom"/>
      <remap from="~imu" to="mavros/imu/data"/>
    </node>

    <node name="blink_processor" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="false"> 
    <!-- <node name="blink_processor" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="debug_roslaunch"> --> 
      <param name="uav_name" type = "string" value="$(arg uav_name)"/>
      <!-- <param name="debug"		type="bool" value="$(arg debug)"/> --> 
      <param name="debug"		type="bool" value="false"/> 
      <param name="visual_debug"		type="bool" value="$(arg visual_debug)"/> 
      <!-- <param name="gui"             type="bool" value="false"/> -->
      <param name="gui"             type="bool" value="true"/>
      <param name="publish_visualization"             type="bool" value="false"/>
      <param name="use_camera_for_visualization"         type="bool" value="true"/>
      <param name="visualization_rate"         type="double" value="2.0"/>

      <!-- <rosparam param="frequencies" subst_value="true"> [$(arg frequency1), $(arg frequency2), $(arg frequency3)] </rosparam> -->
      <rosparam param="signal_ids" subst_value="true"> [$(arg id1), $(arg id2), $(arg id3), $(arg id4), $(arg id5), $(arg id6), $(arg id7), $(arg id8), $(arg id9), $(arg id10), $(arg id11), $(arg id12)] </rosparam>

      <param name="sequence_file"         type="string" value="$(arg sequence_file)"/>

      <param name="blink_process_rate"         type="int" value="$(arg blink_process_rate)"/>

      <param name="accumulator_length"         type="int" value="$(arg accumulator_length)"/>
      <param name="pitch_steps"         type="int" value="$(arg pitch_steps)"/>
      <param name="yaw_steps"         type="int" value="$(arg yaw_steps)"/>
      <param name="max_pixel_shift"         type="int" value="$(arg max_pixel_shift)"/>
      <param name="nullify_radius"         type="int" value="5"/>

      <rosparam param="camera_topics"> ["camera_left", "camera_right", "camera_back"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_left", "points_seen_right", "points_seen_back"] </rosparam>
      <rosparam param="blinkers_seen_topics"> ["blinkers_seen_left", "blinkers_seen_right", "blinkers_seen_back"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimated_framerate_left", "estimated_framerate_right", "estimated_framerate_back"] </rosparam>

      <rosparam param="ami_logging_topics"> ["ami_logging_left", "ami_logging_right", "ami_logging_back"] </rosparam>
      <rosparam param="ami_all_seq_info_topics"> ["ami_all_seq_info_left", "ami_all_seq_info_right", "ami_all_seq_info_back"] </rosparam>

      <remap from="~camera_left" to="/$(arg uav_name)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right" to="/$(arg uav_name)/uvdar_bluefox_right/image_raw"/>
      <remap from="~camera_back" to="/$(arg uav_name)/uvdar_bluefox_back/image_raw"/>
      <remap from="~points_seen_left" to="/$(arg uav_name)/uvdar/points_seen_left"/>
      <remap from="~points_seen_right" to="/$(arg uav_name)/uvdar/points_seen_right"/>
      <remap from="~points_seen_back" to="/$(arg uav_name)/uvdar/points_seen_back"/>
      <remap from="~blinkers_seen_left" to="/$(arg uav_name)/uvdar/blinkers_seen_left"/>
      <remap from="~blinkers_seen_right" to="/$(arg uav_name)/uvdar/blinkers_seen_right"/>
      <remap from="~blinkers_seen_back" to="/$(arg uav_name)/uvdar/blinkers_seen_back"/>
      <remap from="~estimated_framerate_left" to="/$(arg uav_name)/uvdar/estimated_framerate_left"/>
      <remap from="~estimated_framerate_right" to="/$(arg uav_name)/uvdar/estimated_framerate_right"/>
      <remap from="~estimated_framerate_back" to="/$(arg uav_name)/uvdar/estimated_framerate_back"/>
      <remap from="~visualization" to="/$(arg uav_name)/uvdar/blink_visualization/image_raw"/>

      <remap from="~ami_logging_left" to="/$(arg uav_name)/uvdar/ami_logging_left"/>
      <remap from="~ami_logging_right" to="/$(arg uav_name)/uvdar/ami_logging_right"/>
      <remap from="~ami_logging_back" to="/$(arg uav_name)/uvdar/ami_logging_back"/>

      <remap from="~ami_all_seq_info_left" to="/$(arg uav_name)/uvdar/ami_all_seq_info_left"/>
      <remap from="~ami_all_seq_info_right" to="/$(arg uav_name)/uvdar/ami_all_seq_info_right"/>
      <remap from="~ami_all_seq_info_back" to="/$(arg uav_name)/uvdar/ami_all_seq_info_back"/>

    </node>

    <!-- <node name="uvdar_pose_calculator_node" pkg="uvdar_core" type="uvdar_pose_calculator_node" output="screen" launch-prefix="debug_roslaunch"> --> 
    <node name="uvdar_pose_calculator_node" pkg="uvdar_core" type="uvdar_pose_calculator_node" output="screen" > 
        <!-- <node name="uvdar_pose_calculator_node" pkg="uvdar_core" type="uvdar_pose_calculator_node" launch-prefix="valgrind -/-tool=callgrind -/-callgrind-out-file=/home/viktor/callgrind.out -/-instr-atstart=no -/-collect-atstart=yes"> --> 
            <!-- <node name="uvdar_pose_calculator_node" pkg="uvdar_core" type="uvdar_pose_calculator_node" launch-prefix="valgrind -/-tool=callgrind -/-callgrind-out-file=/home/viktor/callgrind.out"> --> 
      <param name="uav_name" type = "string" value="$(arg uav_name)"/>
      <!-- <param name="debug"		type="bool" value="true"/> --> 
      <param name="debug"		type="bool" value="false"/> 
      <!-- <param name="gui"             type="bool" value="false"/> -->
      <param name="gui"             type="bool" value="false"/>
      <param name="publish_visualization"             type="bool" value="false"/>
      <param name="publish_constituents"             type="bool" value="true"/>
      <!-- <param name="publish_visualization"             type="bool" value="false"/> -->
      <!-- <param name="frequencies_per_target" type="int" value="1"/> -->
      <!-- <rosparam param="frequencies" subst_value="true"> [$(arg frequency1), $(arg frequency2), $(arg frequency3)] </rosparam> -->
      <rosparam param="signal_ids" subst_value="true"> [$(arg id1), $(arg id2), $(arg id3), $(arg id4), $(arg id5), $(arg id6), $(arg id7), $(arg id8), $(arg id9), $(arg id10), $(arg id11), $(arg id12)] </rosparam>


      <param name="profiling" type="bool" value="$(arg profiling)"/>

      <param name="quadrotor" type="bool" value="false"/>
      <param name="custom_model" type="bool" value="true"/>
      <param name="model_file" type="string" value="$(arg model_file)"/>
      <param name="beacon" type="bool" value="false"/>

      <rosparam param="blinkers_seen_topics"> ["blinkers_seen_left", "blinkers_seen_right", "blinkers_seen_back"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimated_framerate_left", "estimated_framerate_right", "estimated_framerate_back"] </rosparam>
      <rosparam param="camera_frames" subst_value="true"> ["$(arg uav_name)/uvcam_left", "$(arg uav_name)/uvcam_right", "$(arg uav_name)/uvcam_back"] </rosparam>

      <remap from="~blinkers_seen_left" to="/$(arg uav_name)/uvdar/blinkers_seen_left"/>
      <remap from="~blinkers_seen_right" to="/$(arg uav_name)/uvdar/blinkers_seen_right"/>
      <remap from="~blinkers_seen_back" to="/$(arg uav_name)/uvdar/blinkers_seen_back"/>
      <remap from="~estimated_framerate_left" to="/$(arg uav_name)/uvdar/estimated_framerate_left"/>
      <remap from="~estimated_framerate_right" to="/$(arg uav_name)/uvdar/estimated_framerate_right"/>
      <remap from="~estimated_framerate_back" to="/$(arg uav_name)/uvdar/estimated_framerate_back"/>

      <rosparam param="calib_files" subst_value="True"> ["default", "default", "default"] </rosparam>

      <remap from="~constituentPoses1" to="/$(arg uav_name)/uvdar/constituentPosesL"/>
      <remap from="~constituentPoses2" to="/$(arg uav_name)/uvdar/constituentPosesR"/>
      <remap from="~constituentPoses3" to="/$(arg uav_name)/uvdar/constituentPosesB"/>

      <remap from="~measuredPoses1" to="/$(arg uav_name)/uvdar/measuredPosesL"/>
      <remap from="~measuredPoses2" to="/$(arg uav_name)/uvdar/measuredPosesR"/>
      <remap from="~measuredPoses3" to="/$(arg uav_name)/uvdar/measuredPosesB"/>
    </node>
  </group>


</launch>
